python/postprocessing/modules/jme/jetmetUncertainties.py

```diff
@@ -0,0 +1,408 @@
+    '''Return the best match to ptc in matchCollection,
+    which is the closest ptc in delta R,
+    together with the squared distance dR2 between ptc
+    and the match.'''
+    deltaR2Min = float('+inf')
+    bm = None
+    for match in matchCollection:
+        dR2 = deltaR2(ptc, match)
+        if dR2 < deltaR2Min:
+            deltaR2Min = dR2
+            bm = match
+    return bm, deltaR2Min
+
+def matchObjectCollection(ptcs, matchCollection, deltaRMax = 0.4, filter = lambda x,y : True):
+    pairs = {}
+    if len(ptcs)==0:
+        return pairs
+    if len(matchCollection)==0:
+        return dict( list(zip(ptcs, [None]*len(ptcs))) )
+    dR2Max = deltaRMax ** 2
+    for ptc in ptcs:
+        bm, dr2 = bestMatch( ptc, [ mob for mob in matchCollection if filter(object,mob) ] )
+        if dr2 < dR2Max:
+            pairs[ptc] = bm
+        else:
+            pairs[ptc] = None
+    return pairs
+#----------------------------------------------------------------------------------------------------
+
+def square(x):
+    return x*x
+
+class jetmetUncertaintiesProducer(Module):
+    def __init__(self, globalTag, jesUncertainties = [ "Total" ], jetType = "AK4PFchs"):
+
+        #--------------------------------------------------------------------------------------------
+        # CV: gloabTag and jetType not yet used, as there is no consistent set of txt files for
+        #     JES uncertainties and JER scale factors and uncertainties yet
+        #--------------------------------------------------------------------------------------------
+
+        self.jesUncertainties = jesUncertainties
+
+        # smear jet pT to account for measured difference in JER between data and simulation.
+        self.applyJERCorr = True
+
+        self.jetBranchName = "Jet"
+        self.genJetBranchName = "GenJet"
+        self.metBranchName = "MET"
+        self.rhoBranchName = "fixedGridRhoFastjetAll"
+
+        # read jet energy scale (JES) uncertainties
+        self.jesInputFilePath = os.environ['CMSSW_BASE'] + "/src/PhysicsTools/NanoAODTools/data/jme/"
+        if len(jesUncertainties) == 1 and jesUncertainties[0] == "Total":
+            self.jesUncertaintyInputFileName = "Summer16_23Sep2016V4_MC_Uncertainty_AK4PFchs.txt"
+        else:
+            self.jesUncertaintyInputFileName = "Summer16_23Sep2016V4_MC_UncertaintySources_AK4PFchs.txt"
+
+        # read jet energy resolution (JER) and JER scale factors and uncertainties
+        # (the txt files were downloaded from https://github.com/cms-jet/JRDatabase/tree/master/textFiles/Spring16_25nsV10_MC )
+        self.jerInputFilePath = os.environ['CMSSW_BASE'] + "/src/PhysicsTools/NanoAODTools/data/jme/"
+        self.jerInputFileName = "Spring16_25nsV10_MC_PtResolution_AK4PFchs.txt" 
+        self.jerUncertaintyInputFileName = "Spring16_25nsV10_MC_SF_AK4PFchs.txt"
+
+        # initialize random number generator
+        # (needed for jet pT smearing)
+        self.rnd = ROOT.TRandom3(12345)
```

Comment:
we should not do the smearing with TRandom but rather with pt_reco-pt_gen amplification

Reply:
... I would also factorize the smearing into another file